src/liballoc/boxed.rs

```diff
@@ -462,22 +481,14 @@ impl<T: ?Sized> Box<T> {
     #[unstable(feature = "box_into_raw_non_null", issue = "47336")]
     #[inline]
     pub fn into_raw_non_null(b: Box<T>) -> NonNull<T> {
-        Box::into_unique(b).into()
-    }
-
-    #[unstable(feature = "ptr_internals", issue = "none", reason = "use into_raw_non_null instead")]
-    #[inline]
-    #[doc(hidden)]
-    pub fn into_unique(b: Box<T>) -> Unique<T> {
-        let b = mem::ManuallyDrop::new(b);
-        let mut unique = b.0;
+        let mut b = mem::ManuallyDrop::new(b);
         // Box is kind-of a library type, but recognized as a "unique pointer" by
         // Stacked Borrows.  This function here corresponds to "reborrowing to
         // a raw pointer", but there is no actual reborrow here -- so
         // without some care, the pointer we are returning here still carries
         // the tag of `b`, with `Unique` permission.
         // We round-trip through a mutable reference to avoid that.
```

Comment:
@RalfJung Could you say some more about how and why `Box` is special in Stacked Borrows? Could it be "just" another library type? Are `Vec` or `Rc` for example also special?

Reply:
The *how*: `Box` is basically like `&mut` (except that unlike `&mut` we never add "protectors" for `Box`, but that's a detail).

The *why*: As far as I know, rustc emits `noalias` annotations for `Box`. This demonstrates that it is not "just another library type". Those annotations are incorrect unless our aliasing model (i.e., Stacked Borrows) actually treats `Box` special. Additionally, back when I started with Stacked Borrows, we also added `dereferencable` annotations. Right now [we do not do that any more](https://github.com/rust-lang/rust/pull/66645), but the plan is to start doing that again once LLVM weakened `dereferencable` semantics to mean "dereferencable on entry to the function" (as opposed to "dereferencable throughout the entire execution of the function", which is what it means now).

Long-term, the plan is to not make `Box` special but to make `Unique` special instead. At least, @Gankra told me that that was the plan. ;) But that is blocked on figuring out better how exactly we want it to be special.

Reply:
> Are Vec or Rc for example also special?

No, right now only `Box` is special.

Eventually, when `Unique` is special, that would be inherited by `Vec`. `Vec` is, in fact, the main motivation for making `Unique` special.